[SPARK-17619][SQL] To add support for pattern matching in ArrayContains expression

priyankagargnitk opened the pull request:
## What changes were proposed in this pull request?

This change adds support for pattern matching in arrayContains expression for the string arrays.
For eg.
        a. arrayContains ( Seq ( “\\d\\d\\s-\\s\\d\\d”,  null, "", "pattern"), "12 - 20" ) returns true
        b. arrayContains ( Seq ( "\\d\\d\\s-\\s\\d\\d",  "", "pattern"), "132 - 20" ) ) returns  false
        c. arrayContains ( Seq ( "\\d\\d\\s-\\s\\d\\d",  null, ””, "pattern"), "132 - 20" ) ) returns  null

This change is completely backward compatible.
## How was this patch tested?

Added some more test cases for pattern match use case in the following:
         a. CollectionFunctionsSuite.scala
         b. DataFrameFunctionsSuite.scala
         c. ExpressionToSQLSuite.scala

jira entry for detail: https://issues.apache.org/jira/browse/SPARK-17619

priyankagargnitk committed:
[SPARK-17619][SQL] To add support for pattern matching in ArrayContains Expression.

## What changes were proposed in this pull request?
This change adds support for pattern matching in arrayContains expression for the string arrays.
For eg.
        a. arrayContains ( Seq ( “\\d\\d\\s-\\s\\d\\d”,  null, "", "pattern"), "12 - 20" ) returns true
        b. arrayContains ( Seq ( "\\d\\d\\s-\\s\\d\\d",  "", "pattern"), "132 - 20" ) ) returns  false
        c. arrayContains ( Seq ( "\\d\\d\\s-\\s\\d\\d",  null, ””, "pattern"), "132 - 20" ) ) returns  null

This change is completely backward compatible.

## How was this patch tested?
Added some more test cases for pattern match use case in the following:
         a. CollectionFunctionsSuite.scala
         b. DataFrameFunctionsSuite.scala
         c. ExpressionToSQLSuite.scala

jira entry for detail: https://issues.apache.org/jira/browse/SPARK-17619

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -191,11 +193,15 @@ case class SortArray(base: Expression, ascendingOrder: Expression)
 @ExpressionDescription(
-  usage = "_FUNC_(array, value) - Returns TRUE if the array contains the value.",
-  extended = " > SELECT _FUNC_(array(1, 2, 3), 2);\n true")
+  usage = """_FUNC_(array, value) - Returns TRUE if the array contains the value or
```
The problem is that this changes the behavior of this method and even makes it a little surprising. Before, `..array("Mr.X"), "Mrox"..` didn't match but now it does. Accepting strings and regexes in the same place is inherently ambiguous. I don't know if we'd change the meaning of an existing function like this.

priyankagargnitk replied:
So, in that case we can add one more expression , something like ArrayContainsWithPatternMatch?

priyankagar committed:
Revert "[SPARK-17619][SQL] To add support for pattern matching in ArrayContains Expression."

This reverts commit 69f63092ddcb04ea10186d71d407b4886dea3245.

priyankagar committed:
[SPARK-17619][SQL] To add support for pattern matching in ArrayContains Expression.

## What changes were proposed in this pull request?
This change adds new expression ArrayContainsWithPatternMatch , which does the pattern matching for string types and works in the same way as ArrayContains for all other data types.

For eg.
        a. ArrayContainsWithPatternMatch ( Seq ( “\\d\\d\\s-\\s\\d\\d”,  null, "", "pattern"), "12 - 20" ) returns true
        b. ArrayContainsWithPatternMatch ( Seq ( "\\d\\d\\s-\\s\\d\\d",  "", "pattern"), "132 - 20" ) ) returns  false
        c. ArrayContainsWithPatternMatch ( Seq ( "\\d\\d\\s-\\s\\d\\d",  null, ””, "pattern"), "132 - 20" ) ) returns  null

This change is completely backward compatible.

## How was this patch tested?
Added some more test cases for pattern match use case in the following:
         a. CollectionFunctionsSuite.scala
         b. DataFrameFunctionsSuite.scala
         c. ExpressionToSQLSuite.scala

jira entry for detail: https://issues.apache.org/jira/browse/SPARK-17619

priyankagar committed:
[SPARK-17619][SQL] To add a new expression ArrayContainsWithPatternMatch.

## What changes were proposed in this pull request?
This change adds new expression ArrayContainsWithPatternMatch , which does the pattern matching for string types and works in the same way as ArrayContains for all other data types.

For eg.
        a. ArrayContainsWithPatternMatch ( Seq ( “\\d\\d\\s-\\s\\d\\d”,  null, "", "pattern"), "12 - 20" ) returns true
        b. ArrayContainsWithPatternMatch ( Seq ( "\\d\\d\\s-\\s\\d\\d",  "", "pattern"), "132 - 20" ) ) returns  false
        c. ArrayContainsWithPatternMatch ( Seq ( "\\d\\d\\s-\\s\\d\\d",  null, ””, "pattern"), "132 - 20" ) ) returns  null

This change is completely backward compatible.

## How was this patch tested?
Added some more test cases for pattern match use case in the following:
         a. CollectionFunctionsSuite.scala
         b. DataFrameFunctionsSuite.scala
         c. ExpressionToSQLSuite.scala

jira entry for detail: https://issues.apache.org/jira/browse/SPARK-17619

priyankagar committed:
Merge branch 'array_contains_with_pattern_match' of github.com:priyankagargnitk/spark into array_contains_with_pattern_match

priyankagargnitk commented:
Reverted the previous changes that i did in ArrayContains and now a new expression is added as ArrayContainsWithPatternMatch.
